File: engine/animation.py
import time


class ColorAnimation:
    def __init__(self, colors, speed=0.25):
        self.colors = colors or [(0, 120, 255)]
        self.speed = max(0.01, float(speed))
        self.index = 0
        self.timer = 0.0
        self.current_color = self.colors[0]

    def update(self, dt):
        self.timer += dt

        if self.timer >= self.speed:
            self.timer = 0.0
            self.index = (self.index + 1) % len(self.colors)
            self.current_color = self.colors[self.index]

    def get_color(self):
        return self.current_color


class AnimationController:
    def __init__(self):
        self.animations = {}
        self.current_name = None
        self.current = None

    def add(self, name, animation):
        self.animations[name] = animation

        if self.current is None:
            self.play(name)

    def play(self, name):
        if name == self.current_name:
            return

        if name not in self.animations:
            return

        self.current_name = name
        self.current = self.animations[name]

    def update(self, dt):
        if self.current:
            self.current.update(dt)

    def get_color(self):
        if self.current:
            return self.current.get_color()

        return (0, 120, 255)

File: src/engine/animation.rs
#[derive(Debug, Clone)]
pub struct ColorAnimation {
    pub colors: Vec<(u8, u8, u8)>,
    pub speed: f64,
    pub time: f64,
}

impl ColorAnimation {
    pub fn new(colors: Vec<(u8, u8, u8)>, speed: f64) -> Self {
        Self {
            colors,
            speed,
            time: 0.0,
        }
    }

    pub fn update(&mut self, dt: f64) {
        self.time += dt;
    }

    pub fn color(&self) -> (u8, u8, u8) {
        if self.colors.is_empty() {
            return (255, 255, 255);
        }
        let index = ((self.time / self.speed.max(0.0001)) as usize) % self.colors.len();
        self.colors[index]
    }
}

#[derive(Debug, Clone)]
pub struct AnimationController {
    pub current: String,
    pub animations: std::collections::BTreeMap<String, ColorAnimation>,
}

impl Default for AnimationController {
    fn default() -> Self {
        Self {
            current: "IDLE".to_string(),
            animations: Default::default(),
        }
    }
}

impl AnimationController {
    pub fn add(&mut self, name: &str, animation: ColorAnimation) {
        self.animations.insert(name.to_string(), animation);
    }

    pub fn play(&mut self, name: &str) {
        self.current = name.to_string();
    }

    pub fn update(&mut self, dt: f64) {
        if let Some(animation) = self.animations.get_mut(&self.current) {
            animation.update(dt);
        }
    }

    pub fn get_color(&self) -> (u8, u8, u8) {
        self.animations
            .get(&self.current)
            .map(ColorAnimation::color)
            .unwrap_or((0, 120, 255))
    }
}
